src/smvlang/smv_types.cpp

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************\
+
+Module: SMV Types
+
+Author: Daniel Kroening, dkr@amazon.com
+
+\*******************************************************************/
+
+#include "smv_types.h"
+
+#include <set>
+
+smv_enumeration_typet::smv_enumeration_typet(
+  const std::vector<irep_idt> &element_ids)
+  : typet(ID_smv_enumeration)
+{
+  auto &sub = elements_sub();
+  for(auto &element_id : element_ids)
+    sub.push_back(irept{element_id});
+}
+
+std::vector<irep_idt> smv_enumeration_typet::elements() const
+{
+  std::vector<irep_idt> result;
+
+  auto &elements_sub = find(ID_elements).get_sub();
+
+  result.reserve(elements_sub.size());
+
+  for(auto &e : elements_sub)
+    result.push_back(e.id());
+
+  return result;
+}
+
+bool smv_enumeration_typet::is_subset_of(
+  const smv_enumeration_typet &other) const
+{
+  std::set<irep_idt> other_elements;
+  auto &other_elements_sub = other.find(ID_elements).get_sub();
+
+  for(auto &element : other_elements_sub)
+    other_elements.insert(element.id());
+
+  for(auto &element : elements())
```

Comment:
Thank you for removing the quadratic approach. But I don't think we should use `elements()` here either as `for(auto &e : find(ID_elements).get_sub())` would avoid constructing the vector.